src/enpass_fmt_plug.c

```diff
@@ -152,7 +152,7 @@ static int crypt_all(int *pcount, struct db_salt *salt)
 #endif
 	for (index = 0; index < count; index += batch_size) {
 		unsigned char master[MAX_KEYS_PER_CRYPT][32];
-		unsigned char output[24];
+		unsigned char output[32];
```

Comment:
It was always a buffer overrun, in strip and enpass formats (actually identical code). The code "decrypts 8 bytes from offset 16" but both the old and the new AES code (as well as the one we have in OpenCL, and even the OpenSSL library when not using EVP) would in fact decrypt a whole block, writing 8 bytes past the end of that 24 byte buffer. Not sure why ASan never caught it but luckily it looks like probably only the IV buffer was thrashed, with no consequences.

Actually, pristine mbedTLS doesn't allow CBC sizes that are not a multiple of block size. It would throw an error (that none of our formats would check for) and decrypt nothing. I think that is correct and considered keeping it that way, changing the very few formats (like these two) that had obvious cases of it - but after looking at the ssh format I decided the only way not to risk false negatives would be to add error checking in every single format (or make a very crude abort from the AES code).

So I amended mbedTLS's CBC code to work just like the others (potentially overflowing output buffer) and only then return the error. Bug compatibility... At least in the process I fixed these two bugs (all other code I saw looked safe).

Reply:
I'm glad I asked. So you made a major change to mbedTLS code - this is yet another reason to separate the original mbetTLS code vs. your changes into two commits. And to add source code comments explaining those changes.

Reply:
The CBC size thing is already commented. I'll add a first commit with pristine files.